drivers/SPI.cpp

```diff
@@ -16,97 +16,144 @@
+    struct SPI::spi_peripheral_s *result = NULL;
+    core_util_critical_section_enter();
+    for (uint32_t idx = 0; idx < SPI_COUNT; idx++) {
+        if ((_peripherals[idx].name == name) ||
+            (_peripherals[idx].name == 0)) {
+            result = &_peripherals[idx];
+            break;
+        }
+        // XXX: we may want to ensure that it was previously initialized with the same mosi/miso/sclk/ss pins
+    }
+    if (!or_last && (result != NULL) && (result->name == 0)) {
+        result = NULL;
+    }
+
+    core_util_critical_section_exit();
+    return result;
 }
 
 void SPI::format(int bits, int mode)
+{
+    format(bits, (spi_mode_t)mode, true);
+}
+
+void SPI::format(uint8_t bits, spi_mode_t mode, bool msb_first)
 {
     lock();
     _bits = bits;
     _mode = mode;
+    _msb_first = msb_first;
     // If changing format while you are the owner then just
     // update format, but if owner is changed then even frequency should be
     // updated which is done by acquire.
-    if (_owner == this) {
-        spi_format(&_spi, _bits, _mode, 0);
+    if (_self->owner == this) {
+        spi_format(&_self->spi, _bits, _mode, _msb_first?SPI_BIT_ORDERING_MSB_FIRST:SPI_BIT_ORDERING_LSB_FIRST);
     } else {
         _acquire();
     }
     unlock();
 }
 
-void SPI::frequency(int hz)
+uint32_t SPI::frequency(uint32_t hz)
 {
+    uint32_t actual_hz;
     lock();
     _hz = hz;
     // If changing format while you are the owner then just
     // update frequency, but if owner is changed then even frequency should be
     // updated which is done by acquire.
-    if (_owner == this) {
-        spi_frequency(&_spi, _hz);
+    if (_self->owner == this) {
+        actual_hz = spi_frequency(&_self->spi, _hz);
     } else {
-        _acquire();
+        actual_hz = _acquire();
     }
     unlock();
+    return actual_hz;
 }
 
-SPI *SPI::_owner = NULL;
-SingletonPtr<PlatformMutex> SPI::_mutex;
 
-// ignore the fact there are multiple physical spis, and always update if it wasn't us last
-void SPI::aquire()
+void SPI::acquire()
 {
     lock();
-    if (_owner != this) {
-        spi_format(&_spi, _bits, _mode, 0);
-        spi_frequency(&_spi, _hz);
-        _owner = this;
-    }
+    _acquire();
     unlock();
 }
 
 // Note: Private function with no locking
-void SPI::_acquire()
+uint32_t SPI::_acquire()
 {
-    if (_owner != this) {
-        spi_format(&_spi, _bits, _mode, 0);
-        spi_frequency(&_spi, _hz);
-        _owner = this;
+    uint32_t actual_hz = 0;
+    if (_self->owner != this) {
+        spi_format(&_self->spi, _bits, _mode, _msb_first?SPI_BIT_ORDERING_MSB_FIRST:SPI_BIT_ORDERING_LSB_FIRST);
+        actual_hz = spi_frequency(&_self->spi, _hz);
+        _self->owner = this;
     }
+    return actual_hz;
 }
 
 int SPI::write(int value)
 {
     lock();
     _acquire();
-    int ret = spi_master_write(&_spi, value);
+    uint32_t ret = 0;
+    spi_transfer(&_self->spi, &value, (_bits+7)/8, &ret, (_bits+7)/8, NULL);
```

Comment:
This only works little-endian. If attempting portability you need a 3-way switch using `uint8_t`, 1;`uint16_t`, 2; or `uint32_t`, 4.

Reply:
Although, do we ever care? Not sure mbed OS supports big-endian at all, or have any plans to do so. It's just that the HAL api said something about platform endianness, suggesting coping with either.

Reply:
IMHO this method should be deprecated as it encourages reading symbols per symbol which is really bad for performances.
This fallback is provided for compatibility.

Reply:
I assume this will be second step - another PR (deprecation note for this byte oriented methods) ?

Reply:
Wouldn't disagree with this - it's a convenient short-hand, but other APIs like `Socket` and `FileHandle` only give you the block form. If people really need byte-wise stuff, they can probably wrap a helper up themselves.